web/packages/teleport/src/ironrdp/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+#![allow(clippy::new_without_default)]
+
+#[macro_use]
+extern crate log;
+extern crate byteorder;
+extern crate bytes;
+extern crate console_log;
+extern crate ironrdp_graphics;
+extern crate ironrdp_pdu;
+extern crate ironrdp_session;
+extern crate js_sys;
+extern crate tracing;
+extern crate tracing_subscriber;
+extern crate tracing_web;
+extern crate wasm_bindgen;
+extern crate web_sys;
+
+use ironrdp_graphics::image_processing::PixelFormat;
+use ironrdp_pdu::geometry::{InclusiveRectangle, Rectangle};
+use ironrdp_pdu::write_buf::WriteBuf;
+use ironrdp_session::fast_path::UpdateKind;
+use ironrdp_session::image::DecodedImage;
+use ironrdp_session::{
+    fast_path::Processor as IronRdpFastPathProcessor,
+    fast_path::ProcessorBuilder as IronRdpFastPathProcessorBuilder, ActiveStageOutput,
+};
+use js_sys::Uint8Array;
+use std::convert::TryFrom;
+use wasm_bindgen::{prelude::*, Clamped};
+use web_sys::ImageData;
+
+#[wasm_bindgen]
+pub fn init_wasm_log(log_level: &str) {
+    use tracing::Level;
+    use tracing_subscriber::filter::LevelFilter;
+    use tracing_subscriber::fmt::time::UtcTime;
+    use tracing_subscriber::prelude::*;
+    use tracing_web::MakeConsoleWriter;
+
+    // When the `console_error_panic_hook` feature is enabled, we can call the
+    // `set_panic_hook` function at least once during initialization, and then
+    // we will get better error messages if our code ever panics.
+    //
+    // For more details see
+    // https://github.com/rustwasm/console_error_panic_hook#readme
+    console_error_panic_hook::set_once();
+
+    if let Ok(level) = log_level.parse::<Level>() {
+        let fmt_layer = tracing_subscriber::fmt::layer()
+            .with_ansi(false)
+            .with_timer(UtcTime::rfc_3339()) // std::time is not available in browsers
+            .with_writer(MakeConsoleWriter);
+
+        let level_filter = LevelFilter::from_level(level);
+
+        tracing_subscriber::registry()
+            .with(fmt_layer)
+            .with(level_filter)
+            .init();
+
+        debug!("IronRDP wasm log is ready");
+        // TODO(isaiah): is it possible to set up logging for IronRDP trace logs like so: https://github.com/Devolutions/IronRDP/blob/c71ada5783fee13eea512d5d3d8ac79606716dc5/crates/ironrdp-client/src/main.rs#L47-L78
+    }
+}
+
+/// | message type (29) | data_length uint32 | data []byte |
+///
+/// This type is used in javascript pass raw RDP Server Fast-Path Update PDU data to Rust.
+#[wasm_bindgen]
+pub struct RDPFastPathPDU {
+    data: Uint8Array,
+}
+
+#[wasm_bindgen]
+impl RDPFastPathPDU {
+    #[wasm_bindgen(constructor)]
+    pub fn new(data: Uint8Array) -> Self {
+        Self { data }
+    }
+}
+
+struct RustRDPFastPathPDU {
+    data: Vec<u8>,
+}
+
+impl From<RDPFastPathPDU> for RustRDPFastPathPDU {
+    fn from(js_frame: RDPFastPathPDU) -> Self {
+        Self {
+            data: js_frame.data.to_vec(), // TODO(isaiah): is it possible to avoid copy?
```

Comment:
I've read about `wasm_bindgen` a bit, I don't believe it's possible (nor it would be sane to do) if we want to create and return _an object_, but we should be able to take a `&[u8]` as a parameter in a `#[wasm_bindgen]` function and pass a `Uint8Array` from the javascript side, and that would result in no copy being done. Rust would then enforce the usual lifetime rules, since that slice reference would only be valid for the duration of the function call.

Reply:
Good call: https://rustwasm.github.io/docs/wasm-bindgen/reference/types/number-slices.html#number-slices-u8-i8-u16-i16-u32-i32-u64-i64-f32-and-f64

Reply:
https://github.com/gravitational/teleport/pull/35872